banking/src/main/java/rocks/kata/Account.java

```diff
@@ -0,0 +1,65 @@
+package rocks.kata;
+
+import static rocks.kata.OperationType.*;
+
+import rocks.kata.DatabaseClasses.AccountOperationsDatabase;
+
+public class Account {
+
+    private static long numberOfAccounts;
+
+    private int balance;
+    private long uniqueAccountID;
+    private AccountOperationsDatabase operationsDatabase;
+
+    public Account(AccountOperationsDatabase operationsDatabase) {
+        Account.numberOfAccounts++;
+        this.uniqueAccountID = Account.numberOfAccounts;
+        this.balance = 0;
+        this.operationsDatabase = operationsDatabase;
+        createAccountOperationAndAddToDatabase(OPENING, 0);
+
+    }
+
+    public void withdraw(int amount) throws NotEnoughBalanceException {
+        if (amount <= 0) {
+            throw new IllegalArgumentException("The transaction value must be positive number");
+        }
+        if (balance - amount < 0) {
+            throw new NotEnoughBalanceException("Insufficient funds on the account");
+        }
+        createAccountOperationAndAddToDatabase(WITHDRAWAL, amount);
+    }
+
+    public void deposit(int amount) {
+        if (amount <= 0) {
+            throw new IllegalArgumentException("The transaction value must be positive number");
+        }
+        createAccountOperationAndAddToDatabase(DEPOSIT, amount);
+    }
+
+    public String printStatement() {
+        return operationsDatabase.toString(uniqueAccountID);
+    }
+
+    private void createAccountOperationAndAddToDatabase(OperationType typeOfOperation, int amount) {
```

Comment:
znowu wydaje się to niebezpieczne wielowątkowo - zmiana z dwóch miejsc równocześnie doprowadzi do nieprawidłowych wyników

Reply:
Wydaje mi się, że wystarczy zapewnić utrzymanie zmiennej balance jako tej bezpiecznej wielowątkowo, więc AtomicInteger zamiast np. synchronized calculateBalanceAfterOperation. Przy samym dodawaniu do bazy danych nie interesuje mnie jaka jest kolejność, bo można by pobierać timestampa i po nim sortować żeby np. wyświetlić użytkownikow historię operacji.
Correct me if Im wrong.
P.S. Jako fizyka chemika boli mnie używanie atomic jako synonimu niepodzielności :< Nawet jeśli to historyczna rzecz

Reply:
Nie chodzi konkretnie o tę metodę, tylko o cały mechanizm (metody wykorzystujące tę) - withdraw i deposit sprawdzają warunki i sam `AtomicInteger` nie wystarczy, żeby je obsłużyć poprawnie. Nie mam lepszego przykładu pod ręką, także zerknij na to:
https://stackoverflow.com/questions/72853368/sequence-java-with-semaphores/72857052
i poczytaj o race conditions.

Na pocieszenie pomyśl sobie, że atomowość tutaj to nawiązanie do greckiej etymologii, a nie bezpośrednio do fizyki :p